File: Week_03/largest-rectangle-in-histogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        int res = 0;
        if (heights.size() == 0)    return res;

        heights.push_back(0);   // sentinel，用于弹出处理最后一个元素
        vector<int> index;
        for (int i = 0; i < heights.size(); ++i) {
            while (index.size() > 0 && heights[index.back()] > heights[i]) {
                int h = heights[index.back()];
                index.pop_back();
                int leftbound = index.size() > 0 ? index.back() : -1;
                res = max(res, (i - leftbound - 1) * h);
            }
            index.push_back(i);
        }
        return res;
    }
};

File: Week_01/largest-rectangle-in-histogram.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        # base case
        if len(heights) == 0: return 0

        # add sentinel
        heights.append(0)
        max_area = 0
        idxs = [0]
        for i in range(1, len(heights)):
            while len(idxs) > 0 and heights[i] < heights[idxs[-1]]:
                height = heights[idxs[-1]]
                idxs.pop()
                left_idx = idxs[-1] if len(idxs) > 0 else -1
                max_area = max(max_area, height * (i - left_idx - 1))
            idxs.append(i)
        return max_area
